src/BuiltInTools/BrowserRefresh/WebSocketScriptInjection.js

```diff
@@ -131,20 +131,25 @@ setTimeout(async function () {
     }
 
     let applyFailed = false;
-    deltas.forEach(d => {
-      try {
-        window.Blazor._internal.applyHotReload(d.moduleId, d.metadataDelta, d.ilDelta, d.pdbDelta)
-      } catch (error) {
-        console.warn(error);
-        applyFailed = true;
-      }
-    });
+    if (window.Blazor?._internal?.applyHotReload) {
+      // Only apply hot reload deltas if Blazor has been initialized.
+      // It's possible for Blazor to start after the initial page load, so we don't consider skipping this step
+      // to be a failure. These deltas will get applied later, when Blazor completes initialization.
+      deltas.forEach(d => {
+        try {
+          window.Blazor._internal.applyHotReload(d.moduleId, d.metadataDelta, d.ilDelta, d.pdbDelta)
+        } catch (error) {
+          console.warn(error);
+          applyFailed = true;
+        }
+      }); 
+    }
 
     fetch('/_framework/blazor-hotreload', { method: 'post', headers: { 'content-type': 'application/json' }, body: JSON.stringify(deltas) })
       .then(response => {
         if (response.status == 200) {
           const etag = response.headers['etag'];
-          window.sessionStorage.setItem('blazor-webasssembly-cache', { etag, deltas });
```

Comment:
There was an extra 's' here, meaning we never used the cached hot reload deltas 🙃

Even with this fix, the caching mechanism is broken because session storage [stores its values as strings](https://developer.mozilla.org/en-US/docs/Web/API/Storage/getItem#return_value). So we'll need to stringify the deltas here and parse them when they get read back in.

Reply:
Opened https://github.com/dotnet/sdk/issues/33123 to track this.

Reply:
Good catch. I'm surprised this has lived for so long. I'm now desperately waiting for the next fix which should hopefully give noticeable perf boost.